dual-stack blog (#12253)
istio/istio.io: content/en/blog/2022/experimental-dual-stack/index.md

with a new set of endpoints they can be referred to. The original design was created with customer requirements that
specified client originating IPv4 requests should be proxied over IPv4, and the same for an originating IPv6 to be
satisfied over IPv6 (we will refer to this as native IP family going forward).

## Community Feedback

Much of the feedback provided by the community for the original RFC was to change Envoy to better support dual-stack use cases
internally instead of supporting this within Istio. This has led us to a [new design](https://docs.google.com/document/d/15LP2XHpQ71ODkjCVItGacPgzcn19fsVhyE7ruMGXDyU/edit?usp=sharing)
where we have taken lessons learned as well as feedback and have applied them.

## Current Work

We have worked with the Envoy community to address numerous concerns which is a reason why dual-stack enablement has
taken us a while to implement. We have [matched IP Family for outbound listener](https://github.com/envoyproxy/envoy/issues/16804)
and [supported multiple addresses per listener](https://github.com/envoyproxy/envoy/issues/11184). Alex Xu has also
been working fervently to get long outstanding issues resolve, with the ability for Envoy to

> [bind to multiple addresses](https://github.com/envoyproxy/envoy/issues/9811) while we also work on a

Reviewer: 9811 is duplicated with 21640, and 9811 is the old one, we should keep 21640 and the multiple addresses per listener here https://github.com/envoyproxy/envoy/issues/11184, these are two major features I worked on.

> [smarter way to pick endpoints for dual-stack](https://github.com/envoyproxy/envoy/issues/21640). In total the team

Reviewer: I would suggest mentioning `socket options for multiple addresses in bind config` https://github.com/envoyproxy/envoy/pull/23496 and the upcoming `socket options for multiple addresses in listener`. Those are features required by some people in the envoy community for running dualstack. Although the istio needn't support those directly since the user can use them with istio envoy filter directly, but I see a lot of people are trying to set socket option, that is gonna be an important requirement.

has submitted over a dozen PRs to Envoy and are working on a least a half dozen more to make Envoy adoption of

> dual stack easier for Istio.

Reviewer 2: Besides the issues and PRs, I suggest that we can paste the API description here:
Multiple addresses support : https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener.proto.html?highlight=additional_addresses
BindConfig support: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#config-core-v3-bindconfig

---
title: "Support of Dual Stack Kubernetes Clusters"
description: "Experimental and Alpha support for Dual Stack Kubernetes Clusters."

> publishdate: 2022-09-07T07:00:00-06:00

Author: We are hoping to publish this January 17th

attribution: "Steve Zhang (Intel), Alex Xu (Intel), Iris Ding (Intel), Jacob Delgado (F5), Ying-chun Cai (formerly F5)"
keywords: [dual-stack]
---

Over the past year, both Intel and F5 have collaborated on an effort to bring support for
[Kubernetes Dual-Stack networking](https://kubernetes.io/docs/concepts/services-networking/dual-stack/) to Istio.

# Background

The journey has taken us longer than anticipated and we continue to have work to do. Initially started with a design based
on a reference design implementation from F5. The design lead to an [RFC](https://docs.google.com/document/d/1oT6pmRhOw7AtsldU0-HbfA0zA26j9LYiBD_eepeErsQ/edit?usp=sharing) that
caused us to re-examine our approach. Notably, there were concerns about memory and performance issues that the community wanted
to be addressed before implementation.

## Experimental Dual Stack branch

> While the team was re-evaluating its options we created a new branch, [experimental-dual-stack](https://github.com/istio/istio/tree/experimental-dual-stack),

Reviewer 3: ```suggestion
While the team was re-evaluating its options we created a new branch, [experimental-dual-stack]({{< github_raw >}}/tree/experimental-dual-stack),
```

in a way that we can implement a big impacting feature within Istio without causing regressions. As such, it would be
considered highly experimental to use--unit and integration tests do not currently pass--yet there are still people
that are testing and using this in their environment.

A big part of the original design was to duplicate a lot of the Envoy configuration. Notably, the listeners, clusters and routes
with a new set of endpoints they can be referred to. The original design was created with customer requirements that
specified client originating IPv4 requests should be proxied over IPv4, and the same for an originating IPv6 to be
satisfied over IPv6 (we will refer to this as native IP family going forward).

## Community Feedback

Much of the feedback provided by the community for the original RFC was to change Envoy to better support dual-stack use cases
internally instead of supporting this within Istio. This has led us to a [new design](https://docs.google.com/document/d/15LP2XHpQ71ODkjCVItGacPgzcn19fsVhyE7ruMGXDyU/edit?usp=sharing)
where we have taken lessons learned as well as feedback and have applied them.

> ## Current Work

Author: New section

## Upcoming Istio 1.17

<wait for @zhlsunshine comments>

Author: 1. Gateways work with dual-stack
2. In-mesh services now can be dual stack

Limitations
1. Connecting draining issues
2. Isn't fully native IPv4 downstream and IPv4 upstream (same for IPv6); many users don't care about this

Author: Is there anything you have done for ztunnel?

to Envoy, such as the ability to [enable socket options on multiple addresses](https://github.com/envoyproxy/envoy/pull/23496),
have landed in the upcoming Istio 1.17 release (e.g. [extra source addresses on inbound clusters](https://github.com/istio/istio/pull/41618)).

The Envoy api changes the team has made can be found under their site at [Listener addresses](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener.proto.html?highlight=additional_addresses)
and [bind config](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#config-core-v3-bindconfig).
Making sure we can have proper support at both the downstream and upstream connection for Envoy is important for realizing
dual-stack support.

In total the team has submitted over a dozen PRs to Envoy and are working on a least a half dozen more to make Envoy adoption of
dual stack easier for Istio.

Meanwhile, on the Istio side you can track the progress in [Issue #40394](https://github.com/istio/istio/issues/40394).
Progress has slowed down a bit lately as we continue working with Envoy on various issues, however, we are hoping to get

> this resolved before Istio 1.17 launching in February.

Reviewer 4: c/launching/launches

> It is our hope that for Istio 1.17 we are able to get rudimentary support for dual-stack working.

Author: We will announce a new blog detailing Istio 1.17 dual-stack capability post-launch.

Author: Steve is working with the ambient team

Reviewer 2: Infact,  Iris and Li Chun are working on the redirection of network traffic for ambient mesh.

### Get involved

While we do not have a publicly available version of Istio available for download with dual-stack capabillities, we are

> excited to get this done and ready in Istio 1.17 due early next year.

Reviewer 2: ```suggestion
excited to get this done and ready in Istio 1.17 due early 2023.
```

We would love your feedback and if you are eager to work with us please stop by our slack channel, #dual-stack within
the [Istio Slack]((https://slack.istio.io/).

> _Thank you to the team that has worked on Istio dual-stack!_

Reviewer 2: we may add the github id or provide a personal github homepage link in this section?

> * _Intel: Steve Zhang, Alex Xu, Iris Ding_

Reviewer 2: we may add the github id or provide a personal github homepage link here?

> * _Intel: Steve Zhang, Alex Xu, Iris Ding_

Reviewer 2: ```suggestion
* _Intel: Steve Zhang(github id: zhlsunshine), Alex Xu(github id: soulxu), Iris Ding(github id: irisdingbj)_
```

> * _Intel: Steve Zhang, Alex Xu, Iris Ding_

Reviewer 2: ```suggestion
* _英特尔： [张怀龙](https://github.com/zhlsunshine), [徐贺杰](https://github.com/soulxu), [丁少君](https://github.com/irisdingbj)_
```

> * _Intel: Steve Zhang, Alex Xu, Iris Ding_

Reviewer 2: ```suggestion
* _英特尔： [Steve Zhang](https://github.com/zhlsunshine), [Alex Xu](https://github.com/soulxu), [Iris Ding](https://github.com/irisdingbj)_
```

Reviewer 3: use Intel in en blog？

> * _F5: Jacob Delgado_

Reviewer 2: ```suggestion
* _F5: Jacob Delgado(github id: jacob-delgado)_
```

> * _F5: Jacob Delgado_

Reviewer 2: ```suggestion
* _F5： [Jacob Delgado](https://github.com/jacob-delgado)_
```

> * _F5: Jacob Delgado_

Reviewer 2: ```suggestion
* _F5： [Jacob Delgado](https://github.com/jacob-delgado)_
```

> * _Yingchun Cai (formerly of F5)_

Reviewer 2: ```suggestion
* _Yingchun Cai (formerly of F5) (github id: ycai-aspen)_
```

> * _Yingchun Cai (formerly of F5)_

Reviewer 2: ```suggestion
* _formerly of F5： [蔡迎春](https://github.com/ycai-aspen)_
```

> * _Yingchun Cai (formerly of F5)_

Reviewer 2: ```suggestion
* _formerly of F5： [Yingchun Cai](https://github.com/ycai-aspen)_
```
